sound/soc/intel/boards/sof_sdw_cs42l43.c

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Based on sof_sdw_rt5682.c
+// Copyright (c) 2023 Intel Corporation
+
+/*
+ *  sof_sdw_cs42l43 - Helpers to handle CS42L43 from generic machine driver
+ */
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/input.h>
+#include <sound/jack.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include <sound/cs42l43.h>
+#include <sound/control.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include <sound/soc-dapm.h>
+#include "sof_sdw_common.h"
+
+static const struct snd_soc_dapm_widget sof_widgets[] = {
+	SND_SOC_DAPM_SPK("Speaker", NULL),
+};
+
+static const struct snd_soc_dapm_route sof_map[] = {
+	{ "Speaker", NULL, "cs42l43 AMP1_OUT_P" },
```

Comment:
I don't have a good understanding of how this 'sof_sdw' machine driver might handle the Cohen+Jamerson combination, maybe it's an entirely different machine driver really since it's a mix of SPI + SoundWire IIRC?

Reply:
The configuration is using Cohen's headset and DMIC + Jamerson's speakers. @plbossart In our usage, both Cohen and Jamerson are SoundWire.

Reply:
@charleskeepax Understand. I will add the widgets and the map back. BTW, do you have recommended INPUT/OUTPUT for headset and DMIC? For example, `PDM1_DIN` and `PDM2_DIN` for DMIC.

Reply:
This is what I currently have internally:

```
static const struct snd_soc_dapm_widget sof_widgets[] = {
        SND_SOC_DAPM_HP("Headphone", NULL),
        SND_SOC_DAPM_MIC("Headset Mic", NULL),
        SND_SOC_DAPM_SPK("Left Speaker", NULL),
        SND_SOC_DAPM_SPK("Right Speaker", NULL),
        SND_SOC_DAPM_MIC("Left Mic", NULL),
        SND_SOC_DAPM_MIC("Right Mic", NULL),
};

static const struct snd_soc_dapm_route sof_map[] = {
        { "Headphone", NULL, "cs42l43 AMP3_OUT" },
        { "Headphone", NULL, "cs42l43 AMP4_OUT" },
        { "cs42l43 ADC1_IN1_P", NULL, "Headset Mic" },
        { "cs42l43 ADC1_IN1_N", NULL, "Headset Mic" },
        { "Left Speaker", NULL, "cs42l43 AMP1_OUT_P" },
        { "Left Speaker", NULL, "cs42l43 AMP1_OUT_N" },
        { "Right Speaker", NULL, "cs42l43 AMP2_OUT_P" },
        { "Right Speaker", NULL, "cs42l43 AMP2_OUT_N" },
        { "cs42l43 PDM2_DIN", NULL, "Left Mic" },
        { "cs42l43 PDM2_DIN", NULL, "Right Mic" },
};
```

Although doing separate left and rights might be excessive, bit of a judgement call there, but certainly happy to drop those if you prefer. I guess that is still missing a route for the stereo microphone jack case, but we can probably ignore that for now as its pretty niche.

On the plus side, I think this routing will be pretty standard everyone seems to be sticking with the connections recommended by the apps guys here.

Reply:
In fact thinking about it merging left and rights is probably better since it allows you to have a single pin switch control for both if one wishes.